### ISSUE TITLE
Multiprocessing with array of objects in shared memory

### ISSUE THREAD

**Issue by hbodory** [numpy/numpy#18424]
In this multiprocessing task, a function takes a very large nested list as an input argument, consisting of thousands of lists of lists of lists. This quickly poses memory constraints - meaning the program is aborted. For this reason, I have been trying to make the nested list accessible via shared memory. I am working on a 64-bit Windows machine, using Python 3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)] and Spyder 4.2.1.

In this context, I encountered two obstacles: First, ShareableList() does not take a nested list as an input. Second, when transforming the nested list to a numpy array, the multiprocessing in combination with the shared memory no longer works. The programme gets stuck.

To make things more specific, I provide a minimal example in the attached zipped folder. The code runs with an array of homogenous data types based on the same list size, see the objects nested_list and numpy_array. However, our data have a different form, similar to the objects nested_list1 and numpy_array1, for which memory sharing does not work (this code at the end of the file is currently commented).
[mp_numpy.zip](https://github.com/numpy/numpy/files/5988030/mp_numpy.zip)

**Comment by mattip** [comment]
You might want to try Dask which I think does this kind of thing.

**Comment by sturlamolden** [comment]
TL;DR version: The problem you describe is due to Windows and solved by using Linux.

Long version:

When you work with shared memory, the shared segment will not have the same base address in each of your processes. This means that a data structure made with pointer (e.g. a linked list) will be invalid. Instead you need to work with relative pointers, that is, offsets from the base address (which in ISO C should have the type ```ptrdiff_t```), instead of the absolute address of each node.

When it comes to shared memory with NumPy, I once wrote something like that, but the code is long outdated. The strategy I used was to use named shared memory. The segment was then pickled by its name. Before you continue on this path, I will warn you that shared memory is often a bad idea, as there is a scalability killed called _false sharing_. This is also the reason why parallel computing based on MPI tends to scale better than code based on OpenMP.

Remember that IPC is fast. ```multiprocessing.Pipe``` is e.g. implemented with Unix domain socket on Linux and named pipe on Windows. The overhead of transferring data is hardly more than that of a memcpy in the kernel. Reading data from a shared segment is not likely to be much faster, contrary to common belief.

You might also want to consider if Windows is really the right operating system to use. For scientific computing is often fasr better to use a variety of Unix (e.g. Linux, FreeBSD, or macOS). In particular, the ```os.fork``` system call is implemented using copy-on-write memory. If you have a large data structure, forking in Unix will usually avoid the out-of-memory problem. A 4k memory page will be copied when you write to the page, but not when you read. In effect you will have a shared instance of your data structure. Windows cannot do this, which is one reason why parallel computing in Python is easier with Linux.

**Comment by seberg** [comment]
Closing.  For the simple reason that it doesn't really seem like something NumPy can (or at least should) focus on trying to find a solution for.  I hope Sturla Molden's reply was useful, though.  (Happy to reopen if anyone disagrees.)
